Sources/Basics/SandboxProfile.swift

```diff
@@ -0,0 +1,103 @@
+    public var allowNetwork: Bool
+
+    /// Configures a SandboxProfile for blocking network access and writing to the file system except to specifically
+    /// permitted locations.
+    public init(_ pathAccessRules: [PathAccessRule] = [], allowNetwork: Bool = false) {
+        self.pathAccessRules = pathAccessRules
+        self.allowNetwork = allowNetwork
+    }
+}
+
+extension SandboxProfile {
+    /// Applies the sandbox profile to the given command line (if the platform supports it), and returns the modified
+    /// command line. On platforms that don't support sandboxing, the unmodified command line is returned.
+    public func apply(to command: [String]) -> [String] {
+        #if os(macOS)
+        return ["/usr/bin/sandbox-exec", "-p", self.generateMacOSSandboxProfileString()] + command
+        #else
+        // rdar://40235432, rdar://75636874 tracks implementing sandboxes for other platforms.
+        return command
+        #endif
+    }
+}
+
+// MARK: - macOS
+
+#if os(macOS)
+fileprivate extension SandboxProfile {
+    /// Private function that generates a Darwin sandbox profile suitable for passing to `sandbox-exec(1)`.
+    func generateMacOSSandboxProfileString() -> String {
+        var contents = "(version 1)\n"
+
+        // Deny everything by default.
+        contents += "(deny default)\n"
+
+        // Import the system sandbox profile.
+        contents += "(import \"system.sb\")\n"
+
+        // Allow operations on subprocesses.
+        contents += "(allow process*)\n"
+
+        // Optionally allow network access (inbound and outbound).
+        if allowNetwork {
+            contents += "(system-network)\n"
+            contents += "(allow network*)\n"
+        }
+
+        // Allow reading any file that isn't protected by TCC or permissions (ideally we'd only allow a specific set
+        // of readable locations, and can maybe tighten this in the future).
+        contents += "(allow file-read*)\n"
+
+        // Apply customized rules for specific file system locations. Everything is readonly by default, so we just
+        // either allow or deny writing, in order. Later rules have precedence over earlier rules.
+        for rule in pathAccessRules {
+            switch rule {
+            case .noaccess(let path):
+                contents += "(deny file-* (subpath \(resolveSymlinksAndQuotePath(path))))\n"
+            case .readonly(let path):
+                contents += "(deny file-write* (subpath \(resolveSymlinksAndQuotePath(path))))\n"
+            case .writable(let path):
+                contents += "(allow file-write* (subpath \(resolveSymlinksAndQuotePath(path))))\n"
+            }
+        }
+        return contents
+    }
+
+    /// Private helper function that resolves an AbsolutePath and returns it as a string quoted for use as a subpath
+    /// in a .sb sandbox profile.
+    func resolveSymlinksAndQuotePath(_ path: AbsolutePath) -> String {
```

Comment:
should this be generalized and included in Basics? I think I saw some variation of this elsewhere

Reply:
That's a good point — in this case I kept it largely unchanged after renaming `Sandbox` to `SandboxProfile`, though I did add in the step to also resolve the path since the sandbox needs that.  I can look at how much work it would be to move this down (at least the quoting).  That part is indeed a relatively common operation.

Reply:
Oh actually I remember why I kept it local... quoting itself is relatively common, but strings are always quoted for a particular purpose.  This quoting happens to be similar to shell quoting or source code quoting, but technically it's using whatever the quoting syntax is for the `.sb` file.  So I think in this case it actually is more correct to keep it private to the writer of the `.sb` file (arguably we should have a `DarwinSandboxProfileWriter` that deals with the syntax, that would be even more correct).  It's a bit unfortunate to have the duplication but the idea was to isolate the syntax from other formats.